target_chains/ethereum/contracts/contracts/pyth/Pyth.sol

```diff
@@ -439,71 +449,109 @@ abstract contract Pyth is
         override
         returns (PythStructs.TwapPriceFeed[] memory twapPriceFeeds)
     {
-        // TWAP requires exactly 2 updates - one for the start point and one for the end point
-        // to calculate the time-weighted average price between those two points
+        // TWAP requires exactly 2 updates: one for the start point and one for the end point
         if (updateData.length != 2) {
             revert PythErrors.InvalidUpdateData();
         }
 
         uint requiredFee = getUpdateFee(updateData);
         if (msg.value < requiredFee) revert PythErrors.InsufficientFee();
 
-        unchecked {
-            twapPriceFeeds = new PythStructs.TwapPriceFeed[](priceIds.length);
-            for (uint i = 0; i < updateData.length - 1; i++) {
+        // Process start update data
+        PythStructs.TwapPriceInfo[] memory startTwapPriceInfos;
+        bytes32[] memory startPriceIds;
+        {
+            uint offsetStart;
+            (
+                offsetStart,
+                startTwapPriceInfos,
+                startPriceIds
+            ) = processSingleTwapUpdate(updateData[0]);
+        }
+
+        // Process end update data
+        PythStructs.TwapPriceInfo[] memory endTwapPriceInfos;
+        bytes32[] memory endPriceIds;
+        {
+            uint offsetEnd;
+            (
+                offsetEnd,
+                endTwapPriceInfos,
+                endPriceIds
+            ) = processSingleTwapUpdate(updateData[1]);
+        }
+
+        // Verify that we have the same number of price feeds in start and end updates
+        if (startPriceIds.length != endPriceIds.length) {
+            revert PythErrors.InvalidTwapUpdateDataSet();
+        }
+
+        // Create a mapping to check that every startPriceId has a matching endPriceId
+        // This ensures price feed continuity between start and end points
+        bool[] memory endPriceIdMatched = new bool[](endPriceIds.length);
+        for (uint i = 0; i < startPriceIds.length; i++) {
+            bool foundMatch = false;
+            for (uint j = 0; j < endPriceIds.length; j++) {
                 if (
-                    (updateData[i].length > 4 &&
-                        UnsafeCalldataBytesLib.toUint32(updateData[i], 0) ==
-                        ACCUMULATOR_MAGIC) &&
-                    (updateData[i + 1].length > 4 &&
-                        UnsafeCalldataBytesLib.toUint32(updateData[i + 1], 0) ==
-                        ACCUMULATOR_MAGIC)
+                    startPriceIds[i] == endPriceIds[j] && !endPriceIdMatched[j]
                 ) {
```

Comment:
we dont need this check anymore because we call `processSingleTwapUpdate` and in the function we call `extractUpdateTypeFromAccumulatorHeader` which checks for the `ACCUMULATOR_MAGIC`